public API endpoint to export schedule final shifts

joeyorlando opened the pull request:
# What this PR does

Closes https://github.com/grafana/oncall-private/issues/1632

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

joeyorlando commented on engine/apps/public_api/views/schedules.py:
```diff
@@ -120,3 +128,84 @@ def export(self, request, pk):
+        end_date_field_name = "end_date"
+
+        def _field_is_required(field_name):
+            return Response(f"{field_name} is required", status=status.HTTP_400_BAD_REQUEST)
+
+        def _field_is_invalid_date(field_name):
+            return Response(
+                f"{field_name} is not a valid date, must be in any valid ISO 8601 format",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        def _convert_date(value):
+            if not value:
+                return None
+
+            try:
+                return date.fromisoformat(value)
+            except ValueError:
+                return None
+
+        start_date_str = request.query_params.get(start_date_field_name, None)
+        start_date = _convert_date(start_date_str)
+
+        end_date_str = request.query_params.get(end_date_field_name, None)
+        end_date = _convert_date(end_date_str)
+
+        if start_date_str is None:
+            return _field_is_required(start_date_field_name)
+        elif start_date is None:
+            return _field_is_invalid_date(start_date_field_name)
+        elif end_date_str is None:
+            return _field_is_required(end_date_field_name)
+        elif end_date is None:
+            return _field_is_invalid_date(end_date_field_name)
+
+        if start_date > end_date:
+            return Response(
+                f"{start_date_field_name} must be less than or equal to {end_date_field_name}",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
```
@grafana/grafana-oncall-backend any suggestions on a cleaner way to do this validation? it works well, but I feel like there is a simpler way w/ `django-rest-framework`

Matvey-Kuk commented on docs/sources/oncall-api-reference/schedules.md:
```diff
@@ -197,3 +197,138 @@ curl "{{API_URL}}/api/v1/schedules/SBM7DV7BKFUYU/" \
+    user_on_call_hours[final_shift["user_pk"]] += shift_time_in_seconds / (60 * 60)
+
+with open(OUTPUT_FILE_NAME, "w") as fp:
+  csv_writer = csv.DictWriter(fp, ["user_pk", "hours_on_call"])
```
I guess it will be useful to request user's e-mail here and write it to the file.

joeyorlando replied:
good idea 👍 implemented in [this commit](https://github.com/grafana/oncall/pull/2047/commits/f9accf5b400d0b9ee636386b72391e23fe263446)

vadimkerr commented on engine/apps/public_api/views/schedules.py:
```diff
@@ -120,3 +126,92 @@ def export(self, request, pk):
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        start_date_field_name = "start_date"
+        end_date_field_name = "end_date"
+
+        def _field_is_required(field_name):
+            return Response(f"{field_name} is required", status=status.HTTP_400_BAD_REQUEST)
+
+        def _field_is_invalid_date(field_name):
+            return Response(
+                f"{field_name} is not a valid date, must be in any valid ISO 8601 format",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        def _convert_date(value):
+            if not value:
+                return None
+
+            try:
+                return date.fromisoformat(value)
+            except ValueError:
+                return None
+
+        start_date_str = request.query_params.get(start_date_field_name, None)
+        start_date = _convert_date(start_date_str)
+
+        end_date_str = request.query_params.get(end_date_field_name, None)
+        end_date = _convert_date(end_date_str)
+
+        if start_date_str is None:
+            return _field_is_required(start_date_field_name)
+        elif start_date is None:
+            return _field_is_invalid_date(start_date_field_name)
+        elif end_date_str is None:
+            return _field_is_required(end_date_field_name)
+        elif end_date is None:
+            return _field_is_invalid_date(end_date_field_name)
```
maybe something like this?

```suggestion
        class FinalShiftQueryParamsSerializer(serializers.Serializer):
            start_date = serializers.DateField(required=True)
            end_date = serializers.DateField(required=True)

            def validate(self, attrs):
                if attrs["start_date"] > attrs["end_date"]:
                    raise serializers.ValidationError("start_date must be less than or equal to end_date")

                if attrs["end_date"] - attrs["start_date"] > timezone.timedelta(days=365):
                    raise serializers.ValidationError("The difference between start_date and end_date must be less than one year (365 days)")

                return attrs

        serializer = FinalShiftQueryParamsSerializer(data=request.query_params)
        serializer.is_valid(raise_exception=True)

        start_date = serializer.validated_data["start_date"]
        end_date = serializer.validated_data["end_date"]
```

vadimkerr commented on engine/apps/schedules/models/on_call_schedule.py:
```diff
@@ -97,8 +98,16 @@ class ScheduleEvent(TypedDict):
 
+class ScheduleFinalShift(TypedDict):
+    user_pk: str
+    user_email: str
```
Maybe add a key for username as well?

vadimkerr commented on docs/sources/oncall-api-reference/schedules.md:
```diff
@@ -195,3 +195,159 @@ curl "{{API_URL}}/api/v1/schedules/SBM7DV7BKFUYU/" \
+  "results": [
+    {
+      "user_pk": "UC2CHRT5SD34X",
+      "user_email": "spongebob@grafana.com",
```
Maybe it would be a good idea to use something more neutral like `alice@example.com`, `bob@example.com`, etc.? 
I personally like SpongeBob SquarePants a lot, but probably there are people that don't feel the same way 😄

joeyorlando commented:
@vadimkerr thanks for the useful feedback. Addressed all of your comments in [this commit](https://github.com/grafana/oncall/pull/2047/commits/3d97983dcbc24988ada2ca78f28cfa044f51234c) 👍
